### PR TITLE
Fix update apt hold pkgs

### LINKED ISSUE #30733: version-controlled packages with hold: True can no longer be upgraded by salt
Hi,

A version-controlled package installed like this cannot be further upgraded when the version is upped:
```
jre installation:
  pkg.installed:
    - name: {{ pillar['jvm']['jre_name'] }}
    - version: {{ pillar['jvm']['version'] }}
    - refresh: True
    - hold: True
```

`hold: True` is used to prevent auto-upgrading but should not prevent a forced installation. An excerpt of man apt-mark:
```
hold is used to mark a package as held back, which will prevent the package from being automatically installed, upgraded or removed.
```

The reason is that the command line is incomplete:
```
2016-01-29 11:13:59,409 [salt.loaded.int.module.cmdmod][ERROR   ][25920] Command ['apt-get', '-q', '-y', '-o', 'DPkg::Options::=--force-confold', '-o', 'DPkg::Options::=--force-confdef', 'install', 'openjdk-8-jdk=8u72-b15-1~trusty1'] failed with return code: 100
```

Apt suggests to add --force-yes: `E: There are problems and -y was used without --force-yes`

I tried the `--ignore-hold` option too but it seems to be buggy:
```
https://bugs.launchpad.net/ubuntu/+source/apt/+bug/1226168
```

A solution could be to add `--force-yes` if and only if `hold: True` and `--ignore-hold` when the Apt bug is fixed.

Best,

Erick
```
---------------------
# salt --versions-report
Salt Version:
           Salt: 2015.8.3
 
Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
           Mako: 0.9.1
         PyYAML: 3.10
          PyZMQ: 14.0.1
         Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.4
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 1.5
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.3.0
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: 0.3.6
          smmap: 0.8.2
        timelib: Not Installed
 
System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-24-generic
         system: Ubuntu 14.04 trusty
```

### PR DESCRIPTION
**Pull request by abednarik**
The goal here is to allow updating hold packages, even when hold is set to True.

Fixes #30733.

### CODE REVIEW

**Comment by jtand** [comment]
@abednarik It looks like this caused some test failures. Would you mind taking a look at the Ubuntu failures? Thanks!

**Comment by abednarik** [comment]
Sure  @jtand . Please let me review this. I will let you know when is fixed.

Cheers.

**Comment by abednarik** [comment]
HI @jtand  now tests in Ubuntu looks good. DOn't know why Centos tests are failing but looks something is missing or a misconfiguration.

**Comment by cachedout** [comment]
@terminalmage Could you help review this please?

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
@terminalmage Could you please review this?

File: salt/modules/aptpkg.py [hunk 1]
```diff
@@ -652,6 +654,7 @@ def install(name=None,
         cmds.append(cmd)
 
     if to_reinstall:
+        all_pkgs.extend([x for x in six.itervalues(to_reinstall)])
```
Comment (terminalmage) [inline salt/modules/aptpkg.py]: I don't think this is going to work. Did you test this with a specified version, or with binary packages (i.e. the ``sources`` param)? When a version is specified, The values in ``to_reinstall`` dict will be in the format ``pkgname=version``, or the path to a cached copy of the .deb package if ``sources`` was used. So, the membership check in line 677 below won't find a match in ``all_pkgs``.

This line should probably be changed to:

```python
    all_pkgs.extend(to_reinstall)
```

There should be no need for a list comprehension, as `list.extend(foo)` will extend the list by iterating over ``foo``, and iterating over a dict yields its keys, which in this case would be package names, which is what we want here.

**Comment by abednarik** [comment]
You are right @terminalmage.  I just fix this PR as you describe.

Thanks and sorry for the error.
Cheers.

**Comment by terminalmage** [comment]
Great. Assuming tests pass, I think that this is GTM.

**Comment by blueyed** [comment]
Can this be backported to 2016.3, please?
I'd be happy to provide a PR for it, if it would get accepted.

**Comment by blueyed** [comment]
It should be updated to use the `update_holds` argument that was added for the fix with yum (https://github.com/saltstack/salt/pull/33790) probably?!

**Comment by blueyed** [comment]
There is a new option `--allow-change-held-packages` (which is more specific than the existing `--force-yes`):

    --allow-change-held-packages
           Force yes; this is a dangerous option that will cause apt to
           continue without prompting if it is changing held packages. It
           should not be used except in very special situations. Using it can
           potentially destroy your system! Configuration Item:
           APT::Get::allow-change-held-packages. Introduced in APT 1.1.

However it is only available with APT 1.1.

**Pull request merged**